eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Proposals.hs

```diff
@@ -0,0 +1,88 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Test.Cardano.Ledger.Conway.Proposals where
+
+import Cardano.Ledger.Conway
+import Cardano.Ledger.Conway.Governance
+import Control.DeepSeq (force)
+import Control.Exception (evaluate)
+import Data.Either (isRight)
+import Data.Foldable (foldl', toList)
+import qualified Data.Map.Strict as Map
+import Data.Maybe (fromMaybe)
+import Data.Sequence (fromList)
+import qualified Data.Sequence as Seq
+import Data.Sequence.Strict (StrictSeq)
+import qualified Data.Set as Set
+import Test.Cardano.Ledger.Common
+import Test.Cardano.Ledger.Conway.Arbitrary (
+  ProposalsForEnactment (..),
+  ProposalsNewActions (..),
+ )
+
+takeTail :: StrictSeq (GovActionId c) -> [GovActionId c]
+takeTail = reverse . take 4 . reverse . toList
+
+takeInit :: StrictSeq (GovActionId c) -> [GovActionId c]
+takeInit = take 4 . toList
+
+spec :: Spec
+spec = do
+  describe "Proposals" $ do
+    context "Construction" $ do
+      prop "Adding new nodes keeps Proposals consistent" $
+        \(ProposalsNewActions ps actions :: ProposalsNewActions Conway) ->
+          let ps' =
+                foldl'
+                  (\p action -> fromMaybe (error "Unable to add action") $ proposalsAddAction action p)
+                  ps
+                  actions
+              actionsMap = foldl' (\accum gas -> Map.insert (gasId gas) gas accum) Map.empty actions
+           in actionsMap `shouldBe` (actionsMap `Map.intersection` proposalsActionsMap ps')
+    context "Removal" $ do
+      prop "Removing leaf nodes keeps Proposals consistent" $
+        \(ps :: Proposals Conway) -> do
+          let gais = Set.fromList $ takeTail $ proposalsIds ps
+              ps' = fst $ proposalsRemoveWithDescendants gais ps
+          proposalsSize ps' `shouldBe` proposalsSize ps - Set.size gais
+      prop "Removing root nodes keeps Proposals consistent" $
+        \(ps :: Proposals Conway) -> do
+          let gais = Set.fromList $ takeInit $ proposalsIds ps
+              ps' = fst $ proposalsRemoveWithDescendants gais ps
+          proposalsSize ps' `shouldSatisfy` (<= proposalsSize ps)
+      prop "Removing non-member nodes throws an AssertionFailure" $
+        \(ProposalsNewActions ps actions :: ProposalsNewActions Conway) ->
+          (evaluate . force) (proposalsRemoveWithDescendants (Set.fromList $ gasId <$> actions) ps)
+            `shouldThrow` anyException
```

Comment:
You should almost never expect any exception, cause that could mask some critical bug that throws some other exception that you did not expect.
```suggestion
          evaluate (force (proposalsRemoveWithDescendants (Set.fromList $ gasId <$> actions) ps))
            `shouldThrow` \AssertionFailed {} -> True
```

Reply:
Awesome! Thank you!